### PR TITLE
hmm: Adds h-m-m.

### PR DESCRIPTION
**Pull request by tartavull**
###### Description of changes

h-m-m (pronounced like the interjection "hmm") is a simple, fast, keyboard-centric terminal-based tool for working with mind maps.
https://github.com/nadrad/h-m-m

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/hmm/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      --set PATH ${lib.makeBinPath [
+        php
+        ncurses
+        coreutils
+        xclip
+      ]}
+  '';
+
+  meta = with lib; {
+    description = "h-m-m is a simple, fast, keyboard-centric terminal-based tool for working with mind maps.";
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:35]: This is not to be nit-picky, but it's the type of stuff people with commit access are gonna mention
> Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does.
https://nixos.org/manual/nixpkgs/stable/#var-meta-description
```suggestion
    description = "A simple, fast, keyboard-centric terminal-based tool for working with mind maps";
```

File: pkgs/applications/editors/hmm/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{ stdenv
+, lib
+, makeWrapper
+, php
+, ncurses
+, coreutils
+, xclip
+}:
+
+
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:10]:
```suggestion
```

File: pkgs/applications/editors/hmm/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  '';
+
+  meta = with lib; {
+    description = "h-m-m is a simple, fast, keyboard-centric terminal-based tool for working with mind maps.";
+    homepage = "https://github.com/nadrad/h-m-m";
+    license = licenses.gpl3Only;
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ tartavull ];
+  };
+}
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:41]: You probably need to add this package to pkgs/top-level/all-packages.nix as well, I don't think I see it in this PR.

File: pkgs/applications/editors/hmm/default.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+stdenv.mkDerivation rec {
+  name = "h-m-m";
+
+  src = builtins.fetchGit {
+    url = "https://github.com/nadrad/h-m-m.git";
+    rev = "ecf74433844e5542f78f64f8a79807818743adff";
+  };
+  nativeBuildInputs = [ makeWrapper ];
+
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:19]: It's probably a good idea if you turn all your commits into a single commit called `hmm: init at <version>`

File: pkgs/applications/editors/hmm/default.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  postFixup = ''
+    wrapProgram $out/bin/h-m-m \
+      --set PATH ${lib.makeBinPath [
+        php
+        ncurses
+        coreutils
+        xclip
+      ]}
+  '';
+
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:33]: You likely need to add yourself to the maintainers list. 

https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix

You should do this in a separate commit.

**Review by cafkafk — CHANGES_REQUESTED** [review]
There seem to be two non-trivial issues here:

- You're not in the maintainers file
- This packages is not defined in top-level, so it probably can't be installed.

Otherwise, I just had some nit-picky things I noticed, tbh it's probably a quick fix.

I haven't tested it but ig aside from the suggestions it lgtm :)

**Comment by tartavull** [comment]
@cafkafk Thank you for the handholding on my first contribution!

**Comment by tartavull** [comment]
@cafkafk Do you know how to pass that test?

File: pkgs/applications/editors/hmm/default.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+, xclip
+}:
+
+stdenv.mkDerivation rec {
+  name = "h-m-m";
+
+  src = builtins.fetchGit {
+    url = "https://github.com/nadrad/h-m-m.git";
+    rev = "ecf74433844e5542f78f64f8a79807818743adff";
+  };
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:16]: Looking at the ofborg log https://gist.github.com/GrahamcOfBorg/ed85746086c9a10f1b5e8cc45d350e36
> nix-env failed:
error: access to URI 'https://github.com/nadrad/h-m-m.git' is forbidden in restricted mode
>       … while evaluating the attribute 'src' of the derivation 'h-m-m'

Is there a reason to use fetchGit here instead of 
```suggestion
  src = fetchFromGitHub {
    owner = "nadrad";
    repo = "h-m-m";
    rev = "ecf74433844e5542f78f64f8a79807818743adff";
    sha256 = "sha256-xxxxvT271ASEk1MFZQ/TPdIsDxbGZUCkde4nmKpVQCU=";
  };
```
Here, you would have to change the SHA-256, but it should show what to replace it with when you try to install this package and it fails.

File: pkgs/applications/editors/hmm/default.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp -r h-m-m $out/bin/h-m-m
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:21]: This is just from my understanding of the git repo, I'm not sure this will work at all
```suggestion
    cp $src/h-m-m $out/bin
```
Or maybe this might work wihout the postFixup lines
```nix
    cp $src/h-m-m $out/lib
    echo -e "#!/bin/sh\nphp $out/lib/h-m-m \"$@\"" > $out/bin
```

**Review by cafkafk — CHANGES_REQUESTED** [review]
This is just my initial thoughts, I'm unsure if they will make it work, but they are what I would have tried first.

I'm not sure either will be a fix tbh, if not maybe I'll try to pull your fork and help more actively, just lmk.

File: pkgs/applications/editors/hmm/default.nix [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{ stdenv
+, lib
+, makeWrapper
+, php
+, ncurses
+, coreutils
+, xclip
+}:
```
Comment (cafkafk) [inline pkgs/applications/editors/hmm/default.nix:8]:
```suggestion
, xclip
, fetchFromGitHub
}:
```
You would ofc also need this line to use fetchFromGithub if you choose that route

**Comment by tartavull** [comment]
It seems that the tests have passed. Is there any other channged need it?

**Comment by tartavull** [comment]
:/

File: pkgs/applications/editors/hmm/default.nix [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+stdenv.mkDerivation rec {
+  name = "h-m-m";
+
+  src = fetchFromGitHub {
+    owner = "nadrad";
+    repo = "h-m-m";
+    rev = "ecf74433844e5542f78f64f8a79807818743adff";
+    sha256 = "sha256-xxxxvT271ASEk1MFZQ/TPdIsDxbGZUCkde4nmKpVQCU=";
```
Comment (jonringer) [inline pkgs/applications/editors/hmm/default.nix:18]:
```suggestion
    sha256 = "sha256-k35xld5VA/IloE0EMTP6MpKUfhdSOk8zEtmA8j1ZWs0=";
```

```
error: hash mismatch in fixed-output derivation '/nix/store/1x33yqzz74ixzhw4yc5ci2ky6lwbk5il-source.drv':
         specified: sha256-xxxxvT271ASEk1MFZQ/TPdIsDxbGZUCkde4nmKpVQCU=
            got:    sha256-k35xld5VA/IloE0EMTP6MpKUfhdSOk8zEtmA8j1ZWs0=
```

**Review by cyplo — APPROVED** [review]
lgtm :)

**Comment by tartavull** [comment]
is there a reason this hasn't being merged?

**Review by emilytrau — CHANGES_REQUESTED** [review]
Apologies for the delay. Another package named `hmm` has been added in the mean time. Would you rename the package to `h-m-m` please? I will merge ASAP

**Comment by AndersonTorres** [comment]
This should be reformatted according to contributing guide.
